Disable native snap point in slidescroll

muxin opened the pull request:
Fixes https://github.com/ampproject/amphtml/issues/7670

Tested in ios 10.3 beta and 10.2

camelburrito commented:
tested on other devices? platforms?

camelburrito commented:
Also this is not a permanent fix , this is just a stop gap. So just flip the flag in the JS dont cleanup the code yet. We would need a permanant fix (or may be identify the issue and file a webkit bug that would be fixed soon)

muxin commented:
I just tested the code (disable native snap point) on real ios device and the autoplay iterating rapidly bug is fixed but there are other misbehaviors showing up. @camelburrito you can try out http://yuxichen2.herokuapp.com/examples/test.amp.max.html to see the misbehaviors.

PS, I tried to debug what's wrong with the native snap point but didn't have a clue.

muxin commented:
I'm going to debug slidescroll custom snap and hopefully find a fix for the newly found misbehaviour: if user scrolls too much and reach the end of the next slide, the carousel will jump one more slide suddenly.

muxin commented:
Thanks to @camelburrito I figured out the fix for the sudden slide jump in ios. I will keep testing and commit coming.

muxin commented:
@camelburrito I removed `handleCustomElasticScroll_` and the problem (jumping to one more slide suddenly) is solved. I removed it in both android and ios cases and confirmed the slidescroll still functional in both platforms. Do you think it's ok to remove it?

amphtml-team commented:
No :(
On Fri, Mar 10, 2017 at 3:19 PM Yuxi Chen <notifications@github.com> wrote:

> @camelburrito <https://github.com/camelburrito> I removed
> handleCustomElasticScroll_ and the problem (jumping to one more slide
> suddenly) is solved. I removed it in both android and ios cases and
> confirmed the slidescroll still functional in both platforms. Do you think
> it's ok to remove it?
>
> —
>
>
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ampproject/amphtml/pull/7966#issuecomment-285774256>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ANd2kCs_UbZ2E_wm8TJPNNgrOm_eK9cyks5rkbBfgaJpZM4MSzHT>
> .
>
> --
> You received this message because you are subscribed to the Google Groups
> "amphtml-eng-github" group.
> To unsubscribe from this group and stop receiving emails from it, send an
> email to amphtml-eng-github+unsubscribe@google.com.
> To post to this group, send email to amphtml-eng-github@google.com.
> To view this discussion on the web visit
> https://groups.google.com/a/google.com/d/msgid/amphtml-eng-github/ampproject/amphtml/pull/7966/c285774256%40github.com
> <https://groups.google.com/a/google.com/d/msgid/amphtml-eng-github/ampproject/amphtml/pull/7966/c285774256%40github.com?utm_medium=email&utm_source=footer>
> .
>
-- 
-- Sriram

muxin commented:
@camelburrito I updated the heroku code with current PR code: http://yuxichen2.herokuapp.com/examples/test.amp.max.html, please check the swiping behavior. It seems great, I wonder why can't we remove `handleCustomElasticScroll_`

jridgewell commented:
Ping @camelburrito @muxin, this is a P0.

camelburrito commented:
@jridgewell - we are working on it we have a solution and we are testing it out on lots of devices - taking a lot of time. @muxin Let me know when the device testing is done.

camelburrito commented on extensions/amp-carousel/0.1/slidescroll.js:
```diff
@@ -308,7 +306,7 @@ export class AmpSlideScroll extends BaseSlides {
 
     const currentScrollLeft = this.slidesContainer_./*OK*/scrollLeft;
-    if (!this.hasNativeSnapPoints_) {
+    if (!this.isIos_ && !this.snappingInProgress_) {
```
||

muxin commented:
@camelburrito Testing on devices. But PTAL

muxin commented:
Tested, the android and ios non-beta behaves the same as before, and the ios beta uses custom snap without crazy slide switching. @camelburrito PTAL

camelburrito commented on extensions/amp-carousel/0.1/slidescroll.js:
```diff
@@ -112,6 +114,14 @@ export class AmpSlideScroll extends BaseSlides {
+
+    /** @private {boolean} */
+    this.isIosBeta_ = startsWith(platformFor(this.win).getIosVersionString(),
+        '10.3');
```
you should check if the version is > 10.2

muxin replied:
Right now the 10.3 is the only problematic version, I will create a followup issue for comparing IOS minor version.

muxin replied:
Created https://github.com/ampproject/amphtml/issues/8206

camelburrito commented on extensions/amp-carousel/0.1/slidescroll.js:
```diff
@@ -112,6 +114,14 @@ export class AmpSlideScroll extends BaseSlides {
+    /** @private {boolean} */
+    this.isIosBetaExperimentOn_ = isExperimentOn(this.win,
+        'slidescroll-ios-beta');
+
```
disable-css-snap

muxin replied:
done

camelburrito commented on extensions/amp-carousel/0.1/slidescroll.js:
```diff
@@ -112,6 +114,14 @@ export class AmpSlideScroll extends BaseSlides {
+        'slidescroll-ios-beta');
+
+    /** @private {boolean} */
+    this.isIosBeta_ = startsWith(platformFor(this.win).getIosVersionString(),
```
isCssSnapDisabled_

muxin replied:
done

camelburrito commented on extensions/amp-carousel/0.1/slidescroll.js:
```diff
@@ -139,6 +155,11 @@ export class AmpSlideScroll extends BaseSlides {
 
+    // Snap point is buggy in IOS 10.3 (beta), so it is disabled in beta.
+    if (this.isIosBetaExperimentOn_ && this.isIosBeta_) {
+      this.slidesContainer_.classList.add('-amp-disable-snap-type');
```
-amp-slidescroll-dsable-snap

muxin replied:
done

camelburrito commented on extensions/amp-carousel/0.1/slidescroll.js:
```diff
@@ -158,6 +179,12 @@ export class AmpSlideScroll extends BaseSlides {
+
+      // Snap point is buggy in IOS 10.3 (beta), so it is disabled in beta.
+      if (this.isIosBetaExperimentOn_ && this.isIosBeta_) {
+        slideWrapper.classList.add('-amp-disable-snap-coordinate');
```
use only one class

muxin replied:
done

camelburrito commented on extensions/amp-carousel/0.1/amp-carousel.css:
```diff
@@ -180,4 +180,12 @@ amp-carousel .amp-carousel-button.amp-disabled {
   -webkit-overflow-scrolling: touch !important;
 }
+
+.-amp-disable-snap-type {
```
```css
.-amp-slidescroll-no-snap..-amp-slides-container {
  scroll-snap-type: none !important;
}

.-amp-slidescroll-no-snap.-amp-slide-item {
  scroll-snap-coordinate: none !important;
}
```

muxin replied:
done

camelburrito commented on tools/experiments/experiments.js:
```diff
@@ -266,6 +266,11 @@ const EXPERIMENTS = [
+  {
+    id: 'slidescroll-ios-beta',
+    name: 'Slidescroll IOS beta',
+    cleanupIssue: 'https://github.com/ampproject/amphtml/issues/8195',
```
add a spec and link the issue.

muxin replied:
done

camelburrito commented on tools/experiments/experiments.js:
```diff
@@ -266,6 +266,11 @@ const EXPERIMENTS = [
         'amp-sortable-table/amp-sortable-table.md',
   },
+  {
+    id: 'slidescroll-ios-beta',
```
change the expt name.

muxin replied:
done

camelburrito commented on extensions/amp-carousel/0.1/slidescroll.js:
```diff
@@ -112,6 +114,14 @@ export class AmpSlideScroll extends BaseSlides {
+        'slidescroll-disable-css-snap');
+
+    /** @private {boolean} */
+    this.shouldDisableCssSnap_ = startsWith(
```
bring in the experiment here and use one boolean.

muxin replied:
done
